EU-Bound Shipping Notice: Add Customs tooltips and constraints

ThomazFB opened the pull request:
Closes #9595  

Why
==========
When the EU Shipping scenario is met, the Customs form should behave differently. When there's no validation error, the first custom line must always display a clickable info tooltip to return the EU Banner. And the description validation of every Custom Line should ask for input with three characters or more with a different error message based on the web implementation.

How
==========
Introduces two flags inside the `ShippingLabelCustomsFormItemDetailsViewModel`. The `shouldEnforceEUCustomsDescription` will make sure that every Custom line will ask for a description with at least three characters, and the `isEUTooltipAvailable` flag will be active for the first Custom Line of the list, showing the clickable tooltip.

Screen Capture
==========
https://github.com/woocommerce/woocommerce-ios/assets/5920403/a1d636e0-fd25-403d-8187-3b085424d75c

How to Test
==========
### Scenario 1
1. Open the app with a site containing the Shipping Labels plugin configured
2. Open the order details of an order with the `processing` status
3. Hit the `Create Shipping Label` button
4. Configure the Origin and Destination address that meets the US to EU condition (e. g. US to Austria)
6. Move forward with the Shipping Label creation until you reach the `Customs` section and open the Customs form
7. Verify that the Banner is correctly displayed inside the view
8. Verify that every Custom line requests a description at least three characters long
9. Verify that if there's no validation error, the info tooltip only shows up inside the first Custom Line
10. Verify that the info tooltip is clickable and calls the EU notice banner back


### Scenario 2
1. Open the app with a site containing the Shipping Labels plugin configured
2. Open the order details of an order with the `processing` status
3. Hit the `Create Shipping Label` button
4. Configure the Origin and Destination address that doesn't meet the EU condition (e. g. Canadá to Japan)
6. Move forward with the Shipping Label creation until you reach the `Customs` section and open the Customs form
7. Verify that the EU notice Banner doesn't show up
8. Verify that every Custom line now requests only a non-empty description
9. Verify that the info tooltip never shows up

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ThomazFB committed:
Introduce info tooltip flag in ShippingLabelCustomsFormItemDetails.swift

ThomazFB committed:
Add info tooltip icon inside ShippingLabelCustomsFormItemDetails.swift

ThomazFB committed:
Make ShippingLabelCustomsFormInput.swift build with hardcoded flag

ThomazFB committed:
Adjust info tooltip to just appear in the first custom line

ThomazFB committed:
Introduce toolTip tapped callback in ShippingLabelCustomsFormItemDetails.swift

ThomazFB committed:
Reconfigure info tooltip icon image to be wrapped by a button connected with the tap action

ThomazFB committed:
Add infoTooltip action to ShippingLabelCustomsFormInput.swift

ThomazFB committed:
Connect tooltip tap event with viewModel handling

ThomazFB committed:
Fix hound issues

ThomazFB committed:
Refactor trailing closure to use named parameter for better reading

ThomazFB committed:
Define `isEUShippingScenario` field in ShippingLabelCustomsFormInput

ThomazFB committed:
Merge branch 'trunk' into issue/9595-add-customs-info-icon

ThomazFB committed:
Refactor ShippingLabelCustomsFormListViewModel parameters

ThomazFB committed:
Rename shouldDisplayShippingNotice to isShippingNoticeVisible

ThomazFB committed:
Pass isEUShippingScenario flag to Customs UI components

ThomazFB committed:
Reorganize parameters inside Customs UI elements

ThomazFB committed:
Rework how EU shipping flag is handled inside Customs entities

ThomazFB committed:
Reorganize ShippingLabelCustomsFormItemDetailsViewModel.swift flags for better description validation

ThomazFB committed:
Adjust ShippingLabelCustomsFormItemDetailsViewModel.swift description validation to account for EU scenarios

ThomazFB committed:
Move the description error message control to ShippingLabelCustomsFormItemDetailsViewModel.swift

ThomazFB committed:
Fix spacing and documentation inside ShippingLabelCustomsFormItemDetailsViewModel.swift

rachelmcr commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Customs/ItemDetails/ShippingLabelCustomsFormItemDetailsViewModel.swift:
```diff
@@ -14,6 +14,14 @@ final class ShippingLabelCustomsFormItemDetailsViewModel: ObservableObject {
+    ///
+    private let shouldEnforceEUCustomsDescription: Bool
+
+    /// Flags if the Customs info tooltip should be presented.
+    ///
+    private let isEUTooltipAvailable: Bool
```
I was a little confused by this property until I looked at `canDisplayTooltipInfoIcon`, because I thought it was meant to control whether the tooltip was displayed. Maybe it's worth updating the documentation to something other than "should be presented"? Maybe something like:

> Flags if the Customs info tooltip can be presented for this item.

ThomazFB replied:
True! It looks confusing indeed 😅 I updated the documentation here: https://github.com/woocommerce/woocommerce-ios/pull/9740/commits/27aa1f4dcc6fe5200bf7f5fdcce16ce58a004bb5. Let me know what you think, thanks!

rachelmcr replied:
That's much clearer, thanks!

rachelmcr commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Customs/ShippingLabelCustomsFormInputViewModel.swift:
```diff
@@ -99,7 +103,13 @@ final class ShippingLabelCustomsFormInputViewModel: ObservableObject {
+                      countries: countries,
+                      currency: currency,
+                      shouldEnforceEUCustomsDescription: isEUShippingScenario,
+                      isEUTooltipAvailable: index == 0 && isEUShippingScenario)
```
Should we also check the banner state, and only show the tooltip if the banner was dismissed? It felt a little weird to see it when I hadn't dismissed the banner, because I didn't know what it was signalling. It was especially weird to tap on it and not have anything happen:

https://github.com/woocommerce/woocommerce-ios/assets/8658164/99989880-b9c3-496b-9b0a-b502bacc5910

At the very least, if the tooltip is visible when the banner is also visible, we should probably disable it so it doesn't respond to taps.

ThomazFB replied:
I believe the design idea for the tooltip to be displayed alongside the banner is to create a visual guide map of what the banner is talking about and the area that needs attention, so I think this situation is intentional. Still, I suggest we do this as a separate task because I don't see a straightforward way to share this banner visibility condition to the insides of the Custom item list UI without expanding the flag complexity introduced in this PR.

I am facing this: We signal to the Customs line list if the `EU Shipping Scenario` is active. If yes, every single line will adjust the description validation rule. Meanwhile, only the first line will receive the signal to show the tooltip. If I see the solution correctly, we must update the Custom line list every time the banner visibility changes or share the `@Published` flag containing the banner visibility state to the inside of the Customs line list. 

Is my assumption of possible solutions correct? Does it make sense to do it? Or maybe there's a simpler way I'm not detecting? 🤔

ThomazFB replied:
One other possible solution I just thought here would be to scroll up and highlight the banner when the tooltip is clicked, even when the banner is already visible 🤔

rachelmcr replied:
> I believe the design idea for the tooltip to be displayed alongside the banner is to create a visual guide map of what the banner is talking about and the area that needs attention, so I think this situation is intentional.

Ok, that's fair! In that case, it makes sense to keep the tooltip visible. That also avoids the complexity you mentioned — I can see how that would probably not be worth the effort to handle.

This is a really nice idea, though, maybe as a nice-to-have:

> One other possible solution I just thought here would be to scroll up and highlight the banner when the tooltip is clicked, even when the banner is already visible

That would resolve the confusion I had when I tapped the tooltip while the banner was still visible.

ThomazFB replied:
Awesome! Thank you for sharing your thoughts, @rachelmcr! I created the iOS issue to track this enhancement [here](https://github.com/woocommerce/woocommerce-ios/issues/9776), and also created one for the [Android side](https://github.com/woocommerce/woocommerce-android/issues/9077) too.

rachelmcr reviewed (commented):
This works well for the described scenarios 👍 

I have a suggestion below about the tooltip when the banner isn't dismissed. I also noticed that there are some tests that need to be updated with the `isEUShippingScenario` parameter (the app fails to build for testing). (On that note, could we add some tests for the view model logic here?)

ThomazFB committed:
Fix ShippingLabelCustomsFormInputViewModelTests.swift test build failures

ThomazFB committed:
Fix ShippingLabelCustomsFormItemDetailsViewModelTests.swift test build failures

ThomazFB committed:
Add description validation tests in ShippingLabelCustomsFormItemDetailsViewModelTests.swift

ThomazFB committed:
Add tooltip visibility tests in ShippingLabelCustomsFormItemDetailsViewModelTests.swift

ThomazFB committed:
Updates the `isEUTooltipAvailable` documentation

ThomazFB committed:
Merge branch 'trunk' into issue/9595-add-customs-info-icon

ThomazFB commented:
Hey @rachelmcr, thank you once again for your great review full of relevant insights! 🙇 Sorry for submitting the PR with build issues. I've fixed them and increased the unit test scenarios following the new rules introduced here. I added unit tests in https://github.com/woocommerce/woocommerce-ios/pull/9740/commits/521773f84fcbcd08195ba9a60004cb8214a139fc and https://github.com/woocommerce/woocommerce-ios/pull/9740/commits/b1e025a46d7789ff7921fb0a8746dd58b83a6f3d.

The PR is ready for a second round. Thank you!

rachelmcr reviewed (approved):
Looks good, thanks for adding those tests!
